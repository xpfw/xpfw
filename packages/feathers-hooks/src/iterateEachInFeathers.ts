// import { iterateEachInFind } from "@xpfw/dm"

const iterateEachInFeathers =
(app: any, collection: string | string[],
 query: any,  iterator: (q: any) => Promise<any>, options?: {pageSize?: number}) => {
  // const findMethod = (c: string, q: any) => {
  //   return app.service(c).find({query: q})
  // }
  // return iterateEachInFind(collection, query, findMethod, iterator, options)
}

export default iterateEachInFeathers
