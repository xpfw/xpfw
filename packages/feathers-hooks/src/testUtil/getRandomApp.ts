import express from "@feathersjs/express"
import rest from "@feathersjs/express/rest"
import feathers from "@feathersjs/feathers"
import memdb from "feathers-memory"
import mongoService from "feathers-mongodb"
import { MongoClient } from "mongodb"

const memd: any = memdb
const getRandomApp = async (memoryServiceName: string, useMongo?: boolean) => {
  const f: any = feathers()
  const app: any = express.default(f)
  app.use(express.json())
  app.use(express.urlencoded({ extended: true }))
  let col: any | undefined
  let db: any | undefined
  let c: any
  app.configure(rest())
  if (useMongo) {
    c = await MongoClient.connect(`mongodb://localhost:27017/`)
    db  = c.db("feathersxpfwvalidatetests")
    col = db.collection(`testFor${memoryServiceName}`)
    app.use(memoryServiceName, mongoService({
      Model: col,
      paginate: {default: 10, max: 100}
    }))
  } else {
    app.use(memoryServiceName, memd({paginate: {default: 10, max: 100}}))
  }
  return Promise.resolve({
    app,
    service: app.service(memoryServiceName),
    cleanUpDb: async () => {
      if (useMongo) {
        await col.drop()
        await c.close()
      }
    }
  })
}

export default getRandomApp
