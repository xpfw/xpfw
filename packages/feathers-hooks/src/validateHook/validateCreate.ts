import { Method } from "../globals"
import { ValidateHook } from "../typeDef"
import generalValidateHook from "./generalValidateHook"

const validateCreate: ValidateHook = (form, params) => {
  return generalValidateHook(form, Method.Create, params)
}

export default validateCreate
