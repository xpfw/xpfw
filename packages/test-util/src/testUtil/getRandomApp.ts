import { authenticate, AuthenticationService, JWTStrategy } from "@feathersjs/authentication"
import { hooks, LocalStrategy } from "@feathersjs/authentication-local"
import * as express from "@feathersjs/express"
import * as rest from "@feathersjs/express/rest"
import feathers, { Hook, HookContext } from "@feathersjs/feathers"
import * as sios from "@feathersjs/socketio"
import * as memdb from "feathers-memory"
import * as mongoService from "feathers-mongodb"
import { get, isString } from "lodash"
import { MongoClient } from "mongodb"
import convertIds from "./convertIds"
import emptyPort from "./emptyPort"

const sio: any = sios
const res: any = rest
const mongoServic: any = mongoService
const memd: any = memdb
const promisifyListen = (app: any, port: number) => {
  return new Promise((resolve) => {
    let server: any
    server = app.listen(port, () => {
      resolve(server)
    })
  })
}

const channelName = "authenticated"

const getRandomApp = async (memoryServiceNames: string | string[],
                            useMongo?: boolean, ClientHolder?: any, useRest: boolean = false,
                            options?: {
                              userStore?: any, convertIds?: boolean, dbStore?: any, collections?: string[]
                            }) => {
  const f: any = feathers()

  const app: any = express.default(f)
  app.use(express.json())
  app.use(express.urlencoded({ extended: true }))
  let col: any | undefined
  let db: any | undefined
  let c: any
  app.configure(useRest ? res() : sio())
  app.set("authentication", {
    secret: "test_secret",
    entity: "user",
    service: "users",
    authStrategies: [ "jwt", "local" ],
    jwtOptions: {
      header: { typ: "access" },
      audience: "https://xpfw.github.io",
      issuer: "feathers",
      algorithm: "HS256",
      expiresIn: "1d"
    },
    local: {
      usernameField: "email", passwordField: "password"
    }
  })
  const authService = new AuthenticationService(app)
  authService.register("jwt", new JWTStrategy())
  authService.register("local", new LocalStrategy())
  app.use("/authentication", authService)
  let port = -1
  if (ClientHolder) {
    port = await emptyPort()
  }
  if (isString(memoryServiceNames)) {
    memoryServiceNames = [memoryServiceNames]
  }
  if (useMongo) {
    c = await MongoClient.connect(`mongodb://localhost:27017/`, { useNewUrlParser: true })
    db  = c.db("feathersxpfwvalidatetests" + port)
    for (const memoryServiceName of memoryServiceNames) {
      col = db.collection(`testFor${memoryServiceName}`)
      app.use(memoryServiceName, mongoServic({
        Model: col,
        paginate: {default: 10, max: 100},
        whitelist: ["$regex", "$options"]
      }))
      if (get(options, "convertIds", false)) {
        app.service(memoryServiceName).hooks({
          before: {
            create: [convertIds("_id", false)],
            update: [convertIds("_id", false)],
            patch: [convertIds("_id", false)],
            find: [convertIds("_id", true)]
          }
        })
      }
    }
  } else {
    for (const memoryServiceName of memoryServiceNames) {
      app.use(memoryServiceName, memd({paginate: {default: 10, max: 100}}))
    }
  }
  if (!useRest) {
    for (const memoryServiceName of memoryServiceNames) {
      app.service(memoryServiceName).publish((data: any, context: any) => {
        return app.channel(channelName)
      })
    }
  }
  app.use("users", memd({paginate: {default: 10, max: 100}}))
  app.service("users").hooks({
    // Make sure `password` never gets sent to the client
    after: hooks.protect("password"),
    before: {
      find: [
        authenticate("jwt")
      ],
      create: [
        hooks.hashPassword("password")
      ]
    }
  })

  app.on("login", (payload: any, { connection }: any) => {
    if (connection) {
      app.channel(channelName).join(connection)
    }
  })
  let server: any
  if (ClientHolder) {
    server = await promisifyListen(app, port)
    ClientHolder.connectTo(`http://localhost:${port}`, {
      useRest, userStore: get(options, "userStore"),
      dbStore: get(options, "dbStore"), collections: get(options, "collections")
    })
  }
  return Promise.resolve({
    app,
    port,
    cleanUp: async () => {
      if (useMongo) {
        await db.dropDatabase()
        await c.close()
      }
      if (ClientHolder) {
        ClientHolder.disconnect(useRest)
        server.close()
      }
    }
  })
}

export default getRandomApp
