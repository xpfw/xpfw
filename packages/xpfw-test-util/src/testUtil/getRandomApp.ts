import authentication from "@feathersjs/authentication"
import jwt from "@feathersjs/authentication-jwt"
import local, { hooks } from "@feathersjs/authentication-local"
import * as handler from "@feathersjs/errors/handler"
import * as express from "@feathersjs/express"
import * as rest from "@feathersjs/express/rest"
import feathers, { Application, Service } from "@feathersjs/feathers"
import * as sios from "@feathersjs/socketio"
import * as memdb from "feathers-memory"
import * as mongoService from "feathers-mongodb"
import { get, isString } from "lodash"
import { MongoClient } from "mongodb"
import convertIds from "./convertIds"
import emptyPort from "./emptyPort"

const sio: any = sios
const res: any = rest
const mongoServic: any = mongoService
const memd: any = memdb
const auth: any = authentication
const promisifyListen = (app: any, port: number) => {
  return new Promise((resolve) => {
    const server = app.listen(port, () => {
      resolve(server)
    })
  })
}

const channelName = "authenticated"

const getRandomApp = async (memoryServiceNames: string | string[],
                            useMongo?: boolean, ClientHolder?: any, useRest: boolean = false,
                            options?: {
                              userStore?: any, convertIds?: boolean, dbStore?: any, collections?: string[]
                            }) => {
  const app: any = express.default(feathers())
  app.use(express.json())
  app.use(express.urlencoded({ extended: true }))
  let col: any | undefined
  let db: any | undefined
  let c: any
  app.configure(useRest ? res() : sio())
  app.configure(auth({secret: "myawfulsecret"}))
  app.configure(local())
  app.configure(jwt())
  let port = -1
  if (ClientHolder) {
    port = await emptyPort()
  }
  if (isString(memoryServiceNames)) {
    memoryServiceNames = [memoryServiceNames]
  }
  if (useMongo) {
    c = await MongoClient.connect(`mongodb://localhost:27017/`)
    db  = c.db("feathersxpfwvalidatetests" + port)
    for (const memoryServiceName of memoryServiceNames) {
      col = db.collection(`testFor${memoryServiceName}`)
      app.use(memoryServiceName, mongoServic({
        Model: col,
        paginate: {default: 10, max: 100}
      }))
      if (get(options, "convertIds", false)) {
        app.service(memoryServiceName).hooks({
          before: {
            create: [convertIds("_id", false)],
            update: [convertIds("_id", false)],
            patch: [convertIds("_id", false)],
            find: [convertIds("_id", true)]
          }
        })
      }
    }
  } else {
    for (const memoryServiceName of memoryServiceNames) {
      app.use(memoryServiceName, memd({paginate: {default: 10, max: 100}}))
    }
  }
  if (!useRest) {
    for (const memoryServiceName of memoryServiceNames) {
      app.service(memoryServiceName).publish((data: any, context: any) => {
        return app.channel(channelName)
      })
    }
  }
  app.use("users", memd({paginate: {default: 10, max: 100}}))
  app.service("users").hooks({
    // Make sure `password` never gets sent to the client
    after: hooks.protect("password"),
    before: {
      find: [
        auth.hooks.authenticate("jwt")
      ],
      create: [
        hooks.hashPassword({ passwordField: "password" })
      ]
    }
  })
  app.service("authentication").hooks({
    before: {
     create: [
      // You can chain multiple strategies
      auth.hooks.authenticate(["jwt", "local"])
     ],
     remove: [
      auth.hooks.authenticate("jwt")
     ]
    }
   })

  app.on("login", (payload: any, { connection }: any) => {
    if (connection) {
      app.channel(channelName).join(connection)
    }
  })
  let server: any
  if (ClientHolder) {
    server = await promisifyListen(app, port)
    ClientHolder.connectTo(`http://localhost:${port}`, {
      useRest, userStore: get(options, "userStore"),
      dbStore: get(options, "dbStore"), collections: get(options, "collections")
    })
  }
  return Promise.resolve({
    app,
    port,
    cleanUp: async () => {
      if (useMongo) {
        await db.dropDatabase()
        await c.close()
      }
      if (ClientHolder) {
        ClientHolder.disconnect(useRest)
        server.close()
      }
    }
  })
}

export default getRandomApp
