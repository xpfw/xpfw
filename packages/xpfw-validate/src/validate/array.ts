import { cloneDeep, each, get, isArray, isNil, isNumber, isString } from "lodash"
import { ErrorType, FieldType } from "../globals"
import { IErrorBase, ValidateFunction } from "../typeDef"
import validateAny from "./any"
import checkShouldBeRequired from "./util/checkShouldBeRequired"

const validateArray: ValidateFunction = (value, field, params) => {
  const errors: IErrorBase[] = []
  const validArray = isArray(value)
  if (checkShouldBeRequired(field, params) && !validArray) {
    errors.push({
      type: ErrorType.required
    })
  }
  if (validArray) {
    const min = get(field, "validate.min")
    if (isNumber(min)) {
      if (value.length < min) {
        errors.push({
          type: ErrorType.tooLow
        })
      }
    }
    const max = get(field, "validate.max")
    if (isNumber(max)) {
      if (value.length > max) {
        errors.push({
          type: ErrorType.tooHigh
        })
      }
    }
  }
  if (errors.length > 0) {
    return Promise.reject(errors)
  }
  const promises: Array<Promise<any>> = []
  if (!isNil(field.validate)) {
    const subValidate: any = field.validate
    if (isString(subValidate.type) || isNumber(subValidate.type)) {
      each(value, (date) => {
        const tmpFieldOptions = cloneDeep(field)
        tmpFieldOptions.type = subValidate.type
        if (tmpFieldOptions.type === FieldType.Array || tmpFieldOptions.type === FieldType.Object) {
          tmpFieldOptions.validate = subValidate.validate
        }
        promises.push(validateAny(date, tmpFieldOptions, params))
      })
    }
  }
  return Promise.all(promises).then((results) => {
    return value
  })
}

export default validateArray
