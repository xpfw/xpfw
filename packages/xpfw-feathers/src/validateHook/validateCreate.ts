import { globals } from "@xpfw/validate"
import { ValidateHook } from "../typeDef"
import generalValidateHook from "./generalValidateHook"

const validateCreate: ValidateHook = (form, params) => {
  return generalValidateHook(form, globals.Method.Create, params)
}

export default validateCreate
