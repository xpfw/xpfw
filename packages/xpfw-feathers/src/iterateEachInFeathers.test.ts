import { TestDefs } from "@xpfw/validate"
const FormNumberAndRequiredText = TestDefs.FormNumberAndRequiredText
import iterateEachInFeathers from "./iterateEachInFeathers"
import getRandomApp from "./testUtil/getRandomApp"
import validateCreate from "./validateHook/validateCreate"

test("verify func", async () => {
  const collectionName = "iterateEach"
  const { app, service } = await getRandomApp(collectionName)

  service.hooks({before: {create: [validateCreate(FormNumberAndRequiredText)]}})
  for (let i = 0; i < 160; i++) {
    await service.create({
      myString: "obj", myNum: i * 3 / 2, filtered: "out", of: i * 432, the: ["o", "b", "j", "ct"]
    })
  }
  const found: any = []
  await iterateEachInFeathers(app, collectionName, {}, (res: any) => {
    found.push(res)
    return Promise.resolve()
  })
  expect(found).toMatchSnapshot("iterateAll")
})
