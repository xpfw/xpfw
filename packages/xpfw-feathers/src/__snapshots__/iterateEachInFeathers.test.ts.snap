// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`verify func: iterateAll 1`] = `
Array [
  Object {
    "fromCollection": "iterateEach",
    "id": 0,
    "myNum": 0,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 1,
    "myNum": 1.5,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 2,
    "myNum": 3,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 3,
    "myNum": 4.5,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 4,
    "myNum": 6,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 5,
    "myNum": 7.5,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 6,
    "myNum": 9,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 7,
    "myNum": 10.5,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 8,
    "myNum": 12,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 9,
    "myNum": 13.5,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 10,
    "myNum": 15,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 11,
    "myNum": 16.5,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 12,
    "myNum": 18,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 13,
    "myNum": 19.5,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 14,
    "myNum": 21,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 15,
    "myNum": 22.5,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 16,
    "myNum": 24,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 17,
    "myNum": 25.5,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 18,
    "myNum": 27,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 19,
    "myNum": 28.5,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 20,
    "myNum": 30,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 21,
    "myNum": 31.5,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 22,
    "myNum": 33,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 23,
    "myNum": 34.5,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 24,
    "myNum": 36,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 25,
    "myNum": 37.5,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 26,
    "myNum": 39,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 27,
    "myNum": 40.5,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 28,
    "myNum": 42,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 29,
    "myNum": 43.5,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 30,
    "myNum": 45,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 31,
    "myNum": 46.5,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 32,
    "myNum": 48,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 33,
    "myNum": 49.5,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 34,
    "myNum": 51,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 35,
    "myNum": 52.5,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 36,
    "myNum": 54,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 37,
    "myNum": 55.5,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 38,
    "myNum": 57,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 39,
    "myNum": 58.5,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 40,
    "myNum": 60,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 41,
    "myNum": 61.5,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 42,
    "myNum": 63,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 43,
    "myNum": 64.5,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 44,
    "myNum": 66,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 45,
    "myNum": 67.5,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 46,
    "myNum": 69,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 47,
    "myNum": 70.5,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 48,
    "myNum": 72,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 49,
    "myNum": 73.5,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 50,
    "myNum": 75,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 51,
    "myNum": 76.5,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 52,
    "myNum": 78,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 53,
    "myNum": 79.5,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 54,
    "myNum": 81,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 55,
    "myNum": 82.5,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 56,
    "myNum": 84,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 57,
    "myNum": 85.5,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 58,
    "myNum": 87,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 59,
    "myNum": 88.5,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 60,
    "myNum": 90,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 61,
    "myNum": 91.5,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 62,
    "myNum": 93,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 63,
    "myNum": 94.5,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 64,
    "myNum": 96,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 65,
    "myNum": 97.5,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 66,
    "myNum": 99,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 67,
    "myNum": 100.5,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 68,
    "myNum": 102,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 69,
    "myNum": 103.5,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 70,
    "myNum": 105,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 71,
    "myNum": 106.5,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 72,
    "myNum": 108,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 73,
    "myNum": 109.5,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 74,
    "myNum": 111,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 75,
    "myNum": 112.5,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 76,
    "myNum": 114,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 77,
    "myNum": 115.5,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 78,
    "myNum": 117,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 79,
    "myNum": 118.5,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 80,
    "myNum": 120,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 81,
    "myNum": 121.5,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 82,
    "myNum": 123,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 83,
    "myNum": 124.5,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 84,
    "myNum": 126,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 85,
    "myNum": 127.5,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 86,
    "myNum": 129,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 87,
    "myNum": 130.5,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 88,
    "myNum": 132,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 89,
    "myNum": 133.5,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 90,
    "myNum": 135,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 91,
    "myNum": 136.5,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 92,
    "myNum": 138,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 93,
    "myNum": 139.5,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 94,
    "myNum": 141,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 95,
    "myNum": 142.5,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 96,
    "myNum": 144,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 97,
    "myNum": 145.5,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 98,
    "myNum": 147,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 99,
    "myNum": 148.5,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 100,
    "myNum": 150,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 101,
    "myNum": 151.5,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 102,
    "myNum": 153,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 103,
    "myNum": 154.5,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 104,
    "myNum": 156,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 105,
    "myNum": 157.5,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 106,
    "myNum": 159,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 107,
    "myNum": 160.5,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 108,
    "myNum": 162,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 109,
    "myNum": 163.5,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 110,
    "myNum": 165,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 111,
    "myNum": 166.5,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 112,
    "myNum": 168,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 113,
    "myNum": 169.5,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 114,
    "myNum": 171,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 115,
    "myNum": 172.5,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 116,
    "myNum": 174,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 117,
    "myNum": 175.5,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 118,
    "myNum": 177,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 119,
    "myNum": 178.5,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 120,
    "myNum": 180,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 121,
    "myNum": 181.5,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 122,
    "myNum": 183,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 123,
    "myNum": 184.5,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 124,
    "myNum": 186,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 125,
    "myNum": 187.5,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 126,
    "myNum": 189,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 127,
    "myNum": 190.5,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 128,
    "myNum": 192,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 129,
    "myNum": 193.5,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 130,
    "myNum": 195,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 131,
    "myNum": 196.5,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 132,
    "myNum": 198,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 133,
    "myNum": 199.5,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 134,
    "myNum": 201,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 135,
    "myNum": 202.5,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 136,
    "myNum": 204,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 137,
    "myNum": 205.5,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 138,
    "myNum": 207,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 139,
    "myNum": 208.5,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 140,
    "myNum": 210,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 141,
    "myNum": 211.5,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 142,
    "myNum": 213,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 143,
    "myNum": 214.5,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 144,
    "myNum": 216,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 145,
    "myNum": 217.5,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 146,
    "myNum": 219,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 147,
    "myNum": 220.5,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 148,
    "myNum": 222,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 149,
    "myNum": 223.5,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 150,
    "myNum": 225,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 151,
    "myNum": 226.5,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 152,
    "myNum": 228,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 153,
    "myNum": 229.5,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 154,
    "myNum": 231,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 155,
    "myNum": 232.5,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 156,
    "myNum": 234,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 157,
    "myNum": 235.5,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 158,
    "myNum": 237,
    "myString": "obj",
  },
  Object {
    "fromCollection": "iterateEach",
    "id": 159,
    "myNum": 238.5,
    "myString": "obj",
  },
]
`;
